lib/AST/CompilerPlugin.cpp

```diff
@@ -0,0 +1,274 @@
+//===--- CompilerPlugin.cpp - Compile Plugin Support ----------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Compiler plugin support
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/AST/CompilerPlugin.h"
+
+#include "swift/AST/ASTMangler.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/DiagnosticsFrontend.h"
+#include "swift/Demangling/Demangle.h"
+#include "swift/Demangling/Demangler.h"
+#include "swift/Parse/Lexer.h"
+#include "swift/Subsystems.h"
+#include "llvm/Config/config.h"
+#include "llvm/Support/DynamicLibrary.h"
+#include <cstdlib>
+
+using namespace swift;
+
+#define PLUGIN_SUPPORT_MODULE_NAME "_CompilerPluginSupport"
+#define PLUGIN_PROTOCOL_NAME "_CompilerPlugin"
+#define ALLMACROS_PROPERTY_NAME "allMacros"
+
+#if __clang__
+#define SWIFT_CC __attribute__((swiftcall))
+#define SWIFT_CONTEXT __attribute__((swift_context))
+#endif
+
+namespace {
+struct MetadataArrayRef {
+  const void *const *metadataArray;
+  ptrdiff_t size;
+
+  ArrayRef<const void *> array() const {
+    return ArrayRef<const void *>(metadataArray, (size_t)size);
+  }
+};
+} // end anonymous namespace
+
+#if __clang__ && SWIFT_SWIFT_PARSER
+extern "C" SWIFT_CC
+MetadataArrayRef swift_ASTGen_getMacroTypes(const void *getter);
+#endif
+
+static const void *
+getMacroRegistrationPropertyGetter(llvm::sys::DynamicLibrary library,
+                                   StringRef moduleName,
+                                   ASTContext &ctx) {
+  assert(!moduleName.empty());
+  // TODO: Consider using runtime lookup to get all types that conform to the
+  // macro protocol instead of finding a global `allMacros` property.
+  std::string name;
+  // Build a mangling tree for
```

Comment:
This is... quite clever. I wonder if we could do something simpler by forcing some more structure into the plugins themselves. For example, if `allCases` had a `@_silgen_name("swift_plugin_allMacros")` on it, then the name would be the same for all plugins and we wouldn't have to do mangling tricks at all.

Reply:
Wouldn't multiple plugins would end up with colliding symbols with `@_silgen_name("swift_plugin_allMacros")`?

Reply:
We could definitely enforce naming pattern, e.g. `@_silgen_name("swift_plugin_<module_name>_allMacros")`. But I still believe it's more hassle-free to just let the compiler handle this, even though the building a mangling tree isn't pretty.

Reply:
Interesting. Separately I wonder if `RTLD_LOCAL` is a solution to the duplicate symbols problem.